### PR TITLE
doc: add vCAT documentation

### PR DESCRIPTION
**Pull request by donsheng**
This patch adds user guide and high level design for vCAT

Tracked-On: #5917
Signed-off-by: dongshen <dongsheng.x.zhang@intel.com>

### CODE REVIEW

**Comment by donsheng** [comment]
@dbkinder @fitchbe @NanlinXie   can someone review this PR? thanks.

File: doc/developer-guides/hld/hv-vcat.rst [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ACRN vCAT is built on top of ACRN RDT: ACRN RDT provides a number of physical CAT resources
+(CLOS IDs + cache ways), ACRN vCAT exposes some number of virtual CAT resources to VMs and
+then transparently map them to the assigned physical CAT resources in the ACRN hypervisor;
+VM can take advantage of vCAT to prioritize and partition virtual cache ways for its own tasks.
+
+In current CAT implementation, one CLOS ID corresponds to one IA32_type_MASK_n (type: L2 or L3,
+n ranges from 0 to MAX_CACHE_CLOS_NUM_ENTRIES - 1) MSR and a bit in a capacity bitmask (CBM)
+corresponds to one cache way.
+
+On current generation systems, normally L3 cache is shared by all CPUs on the same socket and
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:18]:
```suggestion
On current generation systems, normally L3 cache is shared by all CPU cores on the same socket and
```
?

File: doc/developer-guides/hld/hv-vcat.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+vCAT refers to the virtualization of Cache Allocation Technology (CAT), one of the
+RDT (Resource Director Technology) technologies.
+
+ACRN vCAT is built on top of ACRN RDT: ACRN RDT provides a number of physical CAT resources
+(CLOS IDs + cache ways), ACRN vCAT exposes some number of virtual CAT resources to VMs and
+then transparently map them to the assigned physical CAT resources in the ACRN hypervisor;
+VM can take advantage of vCAT to prioritize and partition virtual cache ways for its own tasks.
+
+In current CAT implementation, one CLOS ID corresponds to one IA32_type_MASK_n (type: L2 or L3,
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:14]:
```suggestion
In current CAT implementation, one CLOS ID corresponds to one ``IA32_type_MASK_n`` (type: L2 or L3,
```

File: doc/developer-guides/hld/hv-vcat.rst [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+n ranges from 0 to MAX_CACHE_CLOS_NUM_ENTRIES - 1) MSR and a bit in a capacity bitmask (CBM)
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:15]:
```suggestion
n ranges from 0 to ``MAX_CACHE_CLOS_NUM_ENTRIES`` - 1) MSR and a bit in a capacity bitmask (CBM)
```

File: doc/developer-guides/hld/hv-vcat.rst [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+vCAT CLOS IDs assignment, it is currently assumed that all the L2/L3 caches (and therefore all CLOS IDs)
+are system-wide caches shared by all cores in the system, this is done for convenience and to simplify
+the vCAT configuration process. If vCAT is enabled for a VM (abbreviated as vCAT VM), there should not
+be any CLOS ID overlap between a vCAT VM and any other VMs. e.g. the vCAT VM has exclusive use of the
+assigned CLOS IDs.
+When assigning cache ways, however, the VM can be given exclusive, shared, or mixed access to the cache
+ways depending on particular performance needs. For example, use dedicated cache ways for RTVM, and use
+shared cache ways between low priority VMs.
+
+In ACRN, vCAT can be configured via the "VM-Configuration". The vCAT cache ways allocated
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:29]: Are you referring to the ACRN Configuration Tool?

File: doc/developer-guides/hld/hv-vcat.rst [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+For further details on the RDT, refer to the ACRN RDT high-level design :ref:`hv_rdt`.
+
+
+High Level ACRN vCAT Design
+***************************
+
+ACRN CAT virtualization support can be divided into two parts:
+
+- CAT Capability Exposure to Guest
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:40]: When you say Guest, do you mean the User VMs **and** the Service VM? If so, then I would call it `Guest VM`. If the Service VM is not part of this, please use `User VM` instead.

File: doc/developer-guides/hld/hv-vcat.rst [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+- CAT resources (CLOS IDs + cache ways) management
+
+The figure below shows high-level design of vCAT in ACRN:
+
+   .. figure:: images/vcat-hld.png
+      :align: center
+
+CAT Capability Exposure to Guest
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:49]: Same question regarding the use of `Guest`.

File: doc/developer-guides/hld/hv-vcat.rst [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+********************************
+ACRN exposes CAT capability and resource to a guest VM via vCPUID and vMSR, as explained
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:51]: Ditto

File: doc/developer-guides/hld/hv-vcat.rst [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+vMSR
+====
+
+The following CAT MSRs will be virtualized for a vCAT VM:
+
+- IA32_PQR_ASSOC
+- IA32_type_MASK_0 ~ IA32_type_MASK_n
+
+By default, after reset, all CPU cores are assigned to COS 0 and all IA32_type_MASK_n MSRs
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:80]: Typo?
```suggestion
By default, after reset, all CPU cores are assigned to CLOS 0 and all IA32_type_MASK_n MSRs
```

File: doc/developer-guides/hld/hv-vcat.rst [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+       ffs64(value): find the first (least significant) bit set in value and return
+       the index of that bit.
+
+       MSR_IA32_type_MASK_0: 0xD10 for L2, 0xC90 for L3
+
+       vclosid: virtual CLOS ID, always starts from 0
+
+       pclosid: corresponding physical CLOS ID for a given vclosid
+
+       vmsr: virtual msr address, passed to vCAT handlers by the
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:133]:
```suggestion
       vmsr: virtual MSR address, passed to vCAT handlers by the
```

File: doc/developer-guides/hld/hv-vcat.rst [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+       caller functions rdmsr_vmexit_handler()/wrmsr_vmexit_handler()
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:134]:
```suggestion
       caller functions ``rdmsr_vmexit_handler()``/``wrmsr_vmexit_handler()``
```

File: doc/developer-guides/hld/hv-vcat.rst [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+       mask_shift = ffs64(max_pcbm)
+
+       vclosid = vmsr - MSR_IA32_type_MASK_0
+       pclosid = vm_config->pclosids[vclosid]
+
+       pmsr = MSR_IA32_type_MASK_0 + pclosid
+       pcbm = vcbm << mask_shift
+       vcbm = pcbm >> mask_shift
+
+       Where
```
Comment (gvancuts) [inline doc/developer-guides/hld/hv-vcat.rst:115]: We should take this part out of the `code-block`, and use double-backticks to highlight the the code-related items.

File: doc/tutorials/vcat_configuration.rst [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+Enable vCAT Configuration
+#########################
+
+vCAT is built on top of RDT, so to use vCAT we must first enable RDT.
+For details on enabling RDT configuration on ACRN, see :ref:`rdt_configuration`.
+For details on ACRN vCAT high-level design, see :ref:`hv_vcat`.
+
+The vCAT feature is disabled by default in ACRN. You can enable it with these settings:
+
+#. Configure system level features:
```
Comment (gvancuts) [inline doc/tutorials/vcat_configuration.rst:12]: I'm assuming we can do this via the UI. If correct, we should add a note telling the user this is done via the UI. We can keep the XML section below indicating that it's listed here as an FYI to show them how those settings are translated into XML in the scenario file.

File: doc/tutorials/vcat_configuration.rst [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        <FEATURES>
+            <RDT>
+                <RDT_ENABLED>y</RDT_ENABLED>
+                <CDP_ENABLED>n</CDP_ENABLED>
+                <VCAT_ENABLED>y</VCAT_ENABLED>
+                <CLOS_MASK></CLOS_MASK>
+            </RDT>
+        </FEATURES>
+
+#. In each guest VM configuration:
```
Comment (gvancuts) [inline doc/tutorials/vcat_configuration.rst:33]: Same question, does `guest` include Service VM here?

File: doc/tutorials/vcat_configuration.rst [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+   - Edit :option:`vm.guest_flags.guest_flag` and add ``GUEST_FLAG_VCAT_ENABLED``
```
Comment (gvancuts) [inline doc/tutorials/vcat_configuration.rst:35]: Is that the case here too, i.e. that `GUEST_FLAG_VCAT_ENABLED` is a flag that can be set for the Service VM?

**Review by gvancuts — COMMENTED** [review]
Very good set of documents!! A few questions and suggestions.

File: doc/developer-guides/hld/hv-vcat.rst (hunk 8 above)
Comment (amyreye) [reply doc/developer-guides/hld/hv-vcat.rst:80]: Recommend using COS instead of CLOS. TCC team is going to start using COS in their next release, and SDM (Software Developer Manual) uses COS (mostly, there is some inconsistency), for example, "Bitmasks are configured via the IA32_resourceType_MASK_n MSRs, where resourceType indicates a resource type (e.g. “L3” for the L3 cache) and n indicates a COS number." https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3b-part-2-manual.pdf.

File: doc/tutorials/vcat_configuration.rst (hunk 13 above)
Comment (donsheng) [reply doc/tutorials/vcat_configuration.rst:33]: yes, it is a generic term

File: doc/tutorials/vcat_configuration.rst (hunk 14 above)
Comment (donsheng) [reply doc/tutorials/vcat_configuration.rst:35]: yes

File: doc/developer-guides/hld/hv-vcat.rst (hunk 5 above)
Comment (donsheng) [reply doc/developer-guides/hld/hv-vcat.rst:40]: ok, will change to Guest VM

File: doc/developer-guides/hld/hv-vcat.rst (hunk 4 above)
Comment (donsheng) [reply doc/developer-guides/hld/hv-vcat.rst:29]: yes, referring to the vcat configuration

File: doc/tutorials/vcat_configuration.rst (hunk 12 above)
Comment (donsheng) [reply doc/tutorials/vcat_configuration.rst:12]: yes, we should be able to do this via UI, but current UI or 2.8 may not support vcat config until  2.9, etc.

File: doc/developer-guides/hld/hv-vcat.rst (hunk 8 above)
Comment (donsheng) [reply doc/developer-guides/hld/hv-vcat.rst:80]: OK, will change to COS

**Comment by donsheng** [comment]
Fixed all review comments

**Review by dbkinder — COMMENTED** [review]
@gvancuts @donsheng If you're both OK with the technical details, we can merge this and come back with a non-technical (grammar/spelling/formatting) edit in a separate PR.

**Comment by donsheng** [comment]
> @gvancuts @donsheng If you're both OK with the technical details, we can merge this and come back with a non-technical (grammar/spelling/formatting) edit in a separate PR.

You two decide, although I will be vacation, I still have time to work on vcat documentation and complete this task.
I will also need to merge the vcat doc changes into 2.7 branch.
So 2 approaches:
1. merge this PR into master branch, then merge the same change into 2.7 branch
2. merge this PR into master branch, come back with a PR also on master.
Then integrate the changes from master to 2.7 branch.

File: doc/tutorials/vcat_configuration.rst [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+     to enable the vCAT feature on the VM.
+
+   - Edit :option:`vm.clos.vcpu_clos` to assign COS IDs to the VM.
+
+     If ``GUEST_FLAG_VCAT_ENABLED`` is not specified for a VM (abbreviated as RDT VM):
+     ``vcpu_clos`` is per CPU in a VM and it configures each CPU in a VM to a desired COS ID.
+     So the number of vcpu_closes is equal to the number of vCPUs assigned.
+
+     If ``GUEST_FLAG_VCAT_ENABLED`` is specified for a VM (abbreviated as vCAT VM):
```
Comment (gvancuts) [inline doc/tutorials/vcat_configuration.rst:46]: I'm not sure I understand the difference between what you call an RDT VM and vCAT VM, can you expand?

File: doc/tutorials/vcat_configuration.rst [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+     ``vcpu_clos`` is not per CPU anymore; instead, it specifies a list of physical COS IDs (minimum 2)
+     that are assigned to a vCAT VM. The number of vcpu_closes is not necessarily equal to
+     the number of vCPUs assigned, but may be not only greater than the number of vCPUs assigned but
+     less than this number. Each vcpu_clos will be mapped to a virtual COS ID, the first vcpu_clos
```
Comment (gvancuts) [inline doc/tutorials/vcat_configuration.rst:50]: I'm having a hard time parsing this sentence: "but may be not only greater than the number of vCPUs assigned but less than this number". Could you rephrase it?

**Comment by dbkinder** [comment]
> > @gvancuts @donsheng If you're both OK with the technical details, we can merge this and come back with a non-technical (grammar/spelling/formatting) edit in a separate PR.
> 
> You two decide, although I will be vacation, I still have time to work on vcat documentation and complete this task. I will also need to merge the vcat doc changes into 2.7 branch.
> 
> So 2 approaches:
> 
> 1. merge this PR into master branch, then merge the same change into 2.7 branch. Come back with 2 same edit PRs,
>    one on master, another one on 2.7 branch.
> 2. merge this PR into master branch, come back with an edit PR also on master.
>    Then integrate the changes (either 2 or combined) from master to 2.7 branch.

I take care of all the doc changes made on master and add them to the v2.7 branch right before we do the release so you won't need to do that.,  We can just focus on getting master documents looking good!

**Comment by donsheng** [comment]
> > > @gvancuts @donsheng If you're both OK with the technical details, we can merge this and come back with a non-technical (grammar/spelling/formatting) edit in a separate PR.
> > 
> > 
> > You two decide, although I will be vacation, I still have time to work on vcat documentation and complete this task. I will also need to merge the vcat doc changes into 2.7 branch.
> > So 2 approaches:
> > 
> > 1. merge this PR into master branch, then merge the same change into 2.7 branch. Come back with 2 same edit PRs,
> >    one on master, another one on 2.7 branch.
> > 2. merge this PR into master branch, come back with an edit PR also on master.
> >    Then integrate the changes (either 2 or combined) from master to 2.7 branch.
> 
> I take care of all the doc changes made on master and add them to the v2.7 branch right before we do the release so you won't need to do that., We can just focus on getting master documents looking good!

Thanks, that sounds good, if so, I think you are right, we can merge this PR first then come back with an edit PR

File: doc/tutorials/vcat_configuration.rst (hunk 16 above)
Comment (donsheng) [reply doc/tutorials/vcat_configuration.rst:50]: this means if  GUEST_FLAG_VCAT_ENABLED is specified, # of vcpu_clos is not necessarily equal to # of vcpus assigned to this vm anymore, the relationship between them can be equal to, less than, or greater than.

File: doc/tutorials/vcat_configuration.rst (hunk 15 above)
Comment (donsheng) [reply doc/tutorials/vcat_configuration.rst:46]: RDT vm: GUEST_FLAG_VCAT_ENABLED is NOT specified, cat is not exposed to guest VM, acrn hypervisor allocate/init/manage the cat resource on behalf of the guest vm transparently.

vcat vm: GUEST_FLAG_VCAT_ENABLED is specified, cat is exposed to guest vm

**Comment by donsheng** [comment]
Regarding the use of CLOS versus COS, the SDM says we can use it interchangeably,  here in acrn code/tool, we also use the term clos instead of cos such as vcpu_clos, CLOS_MASK, etc. I also checked linux kernel code and intel libpqr (cmt-cat), their code also used CLOS instead of COS.

**Comment by amyreye** [comment]
> Regarding the use of CLOS versus COS, here in acrn code/tool, we use the term clos instead of cos such as vcpu_clos, CLOS_MASK, etc. I also checked linux kernel code and intel libpqr (cmt-cat), their code also used CLOS instead of COS.

@donsheng Thank you for checking. Let's keep CLOS for this release.

**Review by dbkinder — APPROVED** [review]
Thanks!
